browser/brave_vpn/dns/brave_vpn_dns_observer_service.cc

```diff
@@ -0,0 +1,277 @@
+/* Copyright 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/brave_vpn/dns/brave_vpn_dns_observer_service.h"
+
+#include <vector>
+
+#include "base/bind.h"
+#include "base/strings/string_split.h"
+#include "base/strings/string_util.h"
+#include "brave/browser/ui/views/brave_vpn/vpn_notification_dialog_view.h"
+#include "brave/components/brave_vpn/pref_names.h"
+#include "chrome/browser/net/secure_dns_config.h"
+#include "chrome/browser/net/secure_dns_util.h"
+#include "chrome/browser/net/stub_resolver_config_reader.h"
+#include "chrome/browser/net/system_network_context_manager.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ui/browser.h"
+#include "chrome/browser/ui/browser_finder.h"
+#include "chrome/browser/ui/browser_window.h"
+#include "chrome/browser/ui/simple_message_box.h"
+#include "chrome/common/pref_names.h"
+#include "chrome/grit/chromium_strings.h"
+#include "components/country_codes/country_codes.h"
+#include "components/grit/brave_components_strings.h"
+#include "components/policy/policy_constants.h"
+#include "components/prefs/pref_service.h"
+#include "net/dns/dns_config.h"
+#include "net/dns/public/dns_over_https_config.h"
+#include "net/dns/public/doh_provider_entry.h"
+#include "net/dns/public/secure_dns_mode.h"
+#include "ui/base/l10n/l10n_util.h"
+
+namespace secure_dns = chrome_browser_net::secure_dns;
+
+namespace brave_vpn {
+
+namespace {
+const char kBraveVpnDnsProvider[] = "Cloudflare";
+const char kDohServersValue[] = "doh_servers";
+const char kDohModeValue[] = "mode";
+
+void SaveUserDNSConfig(PrefService* prefs, SecureDnsConfig& config) {
+  base::Value::Dict user_dns_config;
+  user_dns_config.Set(kDohModeValue,
+                      SecureDnsConfig::ModeToString(config.mode()));
+  user_dns_config.Set(kDohServersValue, config.doh_servers().ToString());
+  prefs->SetDict(prefs::kBraveVPNUserConfig, std::move(user_dns_config));
+}
+
+void SkipDNSDialog(PrefService* prefs, bool checked) {
+  if (!prefs)
+    return;
+  prefs->SetBoolean(prefs::kBraveVPNShowDNSPolicyWarningDialog, !checked);
+}
+
+bool IsValidDoHTemplates(const std::string& templates) {
+  auto urls_whithout_templates = templates;
+  base::ReplaceSubstringsAfterOffset(&urls_whithout_templates, 0, "{?dns}", "");
+  std::vector<std::string> urls =
+      base::SplitString(urls_whithout_templates, "\n", base::TRIM_WHITESPACE,
+                        base::SPLIT_WANT_NONEMPTY);
+  for (const auto& it : urls) {
+    if (GURL(it).is_valid())
+      return true;
+  }
+  return false;
+}
+
+std::string GetFilteredProvidersForCountry() {
+  namespace secure_dns = chrome_browser_net::secure_dns;
```

Comment:
you already have this here https://github.com/brave/brave-core/pull/15989/files#diff-12f809630cbbcd489be11fbed074059194fceac5237d8f2b68219a19ef73a988R36 and you should not do this inside a method anyway. There are only two usages of `chrome_browser_net::secure_dns` so you're adding more text to alias them than you would just using the full namespace so I would remove both of them.